pkg/cli/tail.go

```diff
@@ -222,24 +254,53 @@ func (c *TailCommand) Run(ctx context.Context, args []string) error {
 	return nil
 }
 
-func (c *TailCommand) tail(ctx context.Context) ([]*loggingpb.LogEntry, error) {
-	var p logPuller
-	if c.testPuller != nil {
-		p = c.testPuller
-	} else {
-		logClient, err := logging.NewClient(ctx)
-		if err != nil {
-			return nil, fmt.Errorf("failed to create logging client: %w", err)
+func (c *TailCommand) streamTail(ctx context.Context, extra []validation.Validator, puller logPuller) error {
+	logCh := make(chan *loggingpb.LogEntry)
+	errCh := make(chan error)
+
+	var failCount int
+	var totalCount int
+
+	go func() {
+		for l := range logCh {
+			totalCount++
+			js, err := protojson.Marshal(l)
+			if err != nil {
+				c.Errf("failed to marshal log to json (InsertId: %q): %w", l.InsertId, err)
+				continue
+			}
+			c.Outf(stripSpaces(string(js)))
+			if c.flagValidate {
+				if err := validation.Validate(string(js), extra...); err != nil {
+					failCount++
+					c.Errf("failed to validate log (InsertId: %q): %w\n", l.InsertId, err)
+				} else {
+					c.Outf("Successfully validated log (InsertId: %q)\n", l.InsertId)
+				}
+			}
 		}
-		p = cloudlogging.NewPuller(ctx, logClient, c.flagScope)
-	}
+	}()
 
-	ls, err := p.Pull(ctx, c.queryFilter(), c.flagMaxNum)
-	if err != nil {
-		return nil, fmt.Errorf("failed to pull logs: %w", err)
-	}
+	go func() {
+		if err := puller.StreamPull(ctx, c.queryFilter(), logCh); err != nil {
+			errCh <- fmt.Errorf("StreamPull failed: %w", err)
+		}
+		close(logCh)
+		close(errCh)
+	}()
 
-	return ls, nil
+	select {
+	case <-ctx.Done():
+		if c.flagValidate {
+			c.Outf("Validation failed for %d logs (out of %d)", failCount, totalCount)
+		}
+		return fmt.Errorf("stream tail validate cancelled: %w", ctx.Err())
```

Comment:
Correct myself, maybe log and return nil instead.

Reply:
I think return an error here is OK, as when the CLI cancel, it will be printed out anyway, have a log showing other information like timestamp and stuff isn't really that useful.